modules/swagger-parser-safe-url-resolver/src/main/java/io/swagger/parser/v3/urlresolver/matchers/UrlPatternMatcher.java

```diff
@@ -0,0 +1,60 @@
+package io.swagger.parser.v3.urlresolver.matchers;
+
+import io.swagger.parser.v3.urlresolver.utils.NetUtils;
+
+import java.net.IDN;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.apache.commons.io.FilenameUtils.wildcardMatch;
+
+public class UrlPatternMatcher {
+
+    private final List<String> patterns;
+
+    public UrlPatternMatcher(List<String> patterns) {
+        this.patterns = new ArrayList<>();
+
+        patterns.forEach(pattern -> {
+            String patternLower = pattern.toLowerCase();
+            String hostAndPort = pattern.contains(":") ? patternLower : patternLower + ":*";
+            String[] split = hostAndPort.split(":");
+            String host = Character.isDigit(split[0].charAt(0)) ? split[0] : IDN.toASCII(split[0], IDN.ALLOW_UNASSIGNED);
+            String port = split.length > 1 ? split[1] : "*";
+
+            // Ignore domains that end in a wildcard
+            if (host.length() > 1 && !NetUtils.isIPv4(host.replace("*", "0")) && host.endsWith("*")) {
+                return;
+            }
+
+            this.patterns.add(String.format("%s:%s", host, port));
+        });
+    }
+
+    public boolean matches(String url) {
+        URL parsed;
+        try {
+            parsed = new URL(url.toLowerCase());
+        } catch (MalformedURLException e) {
+            return false;
+        }
+
+        String host = IDN.toASCII(parsed.getHost(), IDN.ALLOW_UNASSIGNED);
```

Comment:
Does IDN.toASCII handle the wildcards ok? The Javadoc says `IllegalArgumentException - if the input string doesn't conform to RFC 3490 specification`. I really can't think of a realistic attack vector for this so if it's problematic feel free to remove it.

Reply:
so `IDN.toASCII` here is only parsing the end urls submitted by the user and seems to work fine with "incorrect" urls like "https://ex*.com/page" as it is using `IDN.ALLOW_UNASSIGNED`.
This is required to be able to parse emojis as punycode and any other non-standard characters